lib/src/stream_subscription_transformer.dart

```diff
@@ -18,10 +18,14 @@ typedef void _SimpleHandler<T>(StreamSubscription<T> inner);
 /// [StreamSubscription.resume] is called, the corresponding handler is invoked.
 /// By default, handlers just forward to the underlying subscription.
 ///
-/// This guarantees that once the transformed [StreamSubscription] has been
-/// canceled, no further callbacks will be invoked. The [handlePause] and
-/// [handleResume] are invoked regardless of whether the subscription is paused
-/// already or not.
+/// Guarantees that once the transformed [StreamSubscription] has been canceled,
+/// no further callbacks will be invoked. The [handlePause] and [handleResume]
+/// are invoked regardless of whether the subscription is paused already or not.
+///
+/// In order to preserve [StreamSubscription] guarantees, **all callbacks must
+/// synchronously call the corresponding method** on the inner
```

Comment:
The one use-case I see where you don't want to recursively call cancel is that you want to reuse the subscription afterwards. That means that you might overwrite the callbacks or call `pause` *instead* of calling `cancel`.

For that to work safely, this wrapper should no longer be able to resume or set callbacks.
How about not making `_inner` final and setting it to `null` on cancel. Maybe also set the callbacks to `null` first. That should cut any connection to the wrapper (at least if you drop being a `DelegatingStreamSubscription` , and always use `_inner`).